File: src/main/java/cz/linkedlist/http/HttpDownloadTask.java
package cz.linkedlist.http;


import cz.linkedlist.*;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.util.concurrent.ListenableFuture;
import org.springframework.util.concurrent.SuccessCallback;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Martin Macko <https://github.com/LinkedList>.
 */
@RequiredArgsConstructor
public class HttpDownloadTask implements Runnable {

    private static final Logger log = LoggerFactory.getLogger(HttpDownloadTask.class);

    private final DownloadTask task;
    private final TileListingService tileListingService;
    private final TileInfoService infoService;
    private final TileDownloader downloader;

    @Override
    public void run() {
        log.info("Checking for new tiles with UTM code: {}, after date: {}", task.getUtm(), task.getDate());
        Collection<LocalDate> dates = tileListingService.availableDatesAfter(task.getUtm(), task.getDate());
        if(!dates.isEmpty()){
            log.info("Found new tileSet for UTM code: {}, new tilesets: {}", task.getUtm(), dates);
            Collection<TileSet> tiles = dates.stream()
                    .map(localDate -> new TileSet(task.getUtm(), localDate)).collect(Collectors.toList());
            final ListenableFuture<List<TileSet>> future = infoService.downTileInfo(tiles);
            future.addCallback(new HttpDownloadTaskSuccess(task, downloader), throwable -> log.error("There was an error while downloading info", throwable));
        } else {
            log.info("No new tiles found with UTM code: {}, after date: {}", task.getUtm(), task.getDate());
        }
    }

    public static class HttpDownloadTaskSuccess implements SuccessCallback<List<TileSet>> {

        private DownloadTask task;
        private List<TileSet> desiredCloudinessTileSets;
        private TileDownloader downloader;

        public HttpDownloadTaskSuccess(DownloadTask task, TileDownloader downloader) {
            this.task = task;
            this.downloader = downloader;
            this.desiredCloudinessTileSets = new ArrayList<>();
        }

        public HttpDownloadTaskSuccess(DownloadTask task, List<TileSet> desiredCloudinessTileSets) {
            this.task = task;
            this.desiredCloudinessTileSets = desiredCloudinessTileSets;
        }

        @Override
        public void onSuccess(List<TileSet> tileSets) {
            final List<TileSet> tileSetsWithDesiredCloudiness = tileSets.stream()
                    .filter(tileSet -> tileSet.cloudiness() < task.getCloudiness())
                    .collect(Collectors.toList());
            if (tileSetsWithDesiredCloudiness.isEmpty()) {
                log.info("No new tileSet with desired cloudiness");
            } else {
                log.info("Found new tileSets with desired cloudiness:");
                for (TileSet set : tileSetsWithDesiredCloudiness) {
                    log.info("Set: {}, cloudiness: {}", set, set.cloudiness());
                }

                desiredCloudinessTileSets.addAll(tileSetsWithDesiredCloudiness);
                if(downloader != null) {
                    tileSetsWithDesiredCloudiness.forEach(tileSet -> downloader.downContent(tileSet, task.getContents()));
                }
            }
        }
    }
}

File: src/main/kotlin/cz/linkedlist/http/HttpDownloadTask.kt
package cz.linkedlist.http


import cz.linkedlist.*
import org.slf4j.LoggerFactory
import org.springframework.util.concurrent.FailureCallback
import org.springframework.util.concurrent.SuccessCallback

/**
 * @author Martin Macko <https:></https:>//github.com/LinkedList>.
 */
class HttpDownloadTask(
    val task: DownloadTask? = null,
    val tileListingService: TileListingService? = null,
    val infoService: TileInfoService? = null,
    val downloader: TileDownloader? = null
) : Runnable {


    override fun run() {
        log.info("Checking for new tiles with UTM code: {}, after date: {}", task!!.utm, task.date)
        val dates = tileListingService!!.availableDatesAfter(task.utm!!, task.date!!)
        if (!dates.isEmpty()) {
            log.info("Found new tileSet for UTM code: {}, new tilesets: {}", task.utm, dates)
            val tiles = dates.map { localDate -> TileSet(task.utm, localDate) }
            val future = infoService!!.downTileInfo(tiles)
            future.addCallback(
                HttpDownloadTaskSuccess(task, downloader!!),
                FailureCallback { throwable -> log.error("There was an error while downloading info", throwable) }
                )
        } else {
            log.info("No new tiles found with UTM code: {}, after date: {}", task.utm, task.date)
        }
    }

    class HttpDownloadTaskSuccess(
        var task: DownloadTask? = null,
        val downloader: TileDownloader? = null,
        var desiredCloudinessTileSets: MutableList<TileSet> = mutableListOf()
    ) : SuccessCallback<List<TileSet>> {

        override fun onSuccess(tileSets: List<TileSet>) {
            val tileSetsWithDesiredCloudiness = tileSets
                .filter { it.cloudiness()!! < task!!.cloudiness!! }
            if (tileSetsWithDesiredCloudiness.isEmpty()) {
                log.info("No new tileSet with desired cloudiness")
            } else {
                log.info("Found new tileSets with desired cloudiness:")
                for (set in tileSetsWithDesiredCloudiness) {
                    log.info("Set: {}, cloudiness: {}", set, set.cloudiness())
                }

                desiredCloudinessTileSets!!.addAll(tileSetsWithDesiredCloudiness)
                if (downloader != null) {
                    tileSetsWithDesiredCloudiness.forEach { tileSet ->
                        downloader.downContent(
                            tileSet,
                            task!!.contents
                        )
                    }
                }
            }
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger(HttpDownloadTask::class.java!!)
    }
}
